File: solutions/Golang/sqrtx_binary_search.go
package main

import "fmt"

func sqrtxBinarySearch(x int) int {
	if x < 2 {
		return x
	}

	left, right := 2, x/2
	for left <= right {
		mid := (left + right) / 2
		if mid*mid == x {
			return mid
		} else if mid*mid < x {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return right
}

func main() {
	fmt.Println(sqrtxBinarySearch(4))  // 2
	fmt.Println(sqrtxBinarySearch(8))  // 2
}

File: solutions/C++/sqrtx_binary_search.cpp
using namespace std;

int sqrtxBinarySearch(int x) {
    if (x < 2) return x;

    long long left = 2, right = x / 2;
    while (left <= right) {
        long long mid = (left + right) / 2;
        if (mid * mid == x) {
            return mid;
        } else if (mid * mid < x) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }

    return right;
}

int main() {
    cout << sqrtxBinarySearch(4) << endl;  // 2
    cout << sqrtxBinarySearch(8) << endl;  // 2
    return 0;
}
